src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,241 @@
+    /// surprising. It can also lead to deadlocks if the match scrutinee uses a function call that
+    /// returns a MutexGuard and then tries to lock again in one of the match arms, because the
+    /// MutexGuard will not be dropped until the end of the match block and thus will not unlock.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     match mutex.lock().unwrap().foo() {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar(); // Deadlock!
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     let is_foo = mutex.lock().unwrap().foo();
+    ///     match is_foo {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar();
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "warns when a temporary of a type with a drop with a significant side-effect might have a surprising lifetime"
+}
+
+declare_lint_pass!(SignificantDropInScrutinee => [SIGNIFICANT_DROP_IN_SCRUTINEE]);
+
+impl<'tcx> LateLintPass<'tcx> for SignificantDropInScrutinee {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let Some((sig_temp_span, expr_span)) = has_significant_drop_in_scrutinee(cx, expr) {
+            let replacement = reindent_multiline(
+                Cow::Owned(format!("let value = {};\nmatch ", snippet(cx, expr_span, ".."))),
+                true,
+                indent_of(cx, expr_span)
+            );
+
+            span_lint_and_then(
+                cx,
+                SIGNIFICANT_DROP_IN_SCRUTINEE,
+                sig_temp_span,
+                "temporary with significant drop in match scrutinee",
+                |diag| {
+                    diag.multipart_suggestion(
+                        "try getting the temporary outside the match",
+                        vec![
+                            (expr.span.until(expr_span), replacement.to_string()),
+                            (expr_span, "value".to_owned()),
```

Comment:
So after the discussion on Zulip getting the span with `expr.shrink_to_lo()` and then suggesting to insert `let value = ...;\n` may just work. So:

```suggestion
                            (expr.span.shrink_to_lo(), replacement.to_string()),
                            (expr_span, "value".to_owned()),
```

and removing the "match" from the `replacement` string.

Reply:
Thank you for the suggestion/catching the bug! This works a lot nicer now.

Something I was a little surprised by: I had to add trailing spaces to get the `match` line to have the correct indent. I'm not sure if there's a better way to do this though.

Reply:
According to Esteban on Zulip the way you implemented this is the best way to do that. :+1: